src/storage.rs

```diff
@@ -2226,27 +2230,21 @@ impl Cache {
         guard.command_history.last_fetch = Some(last_fetch);
     }
 
-    /// Synchronous routine to get whatever publisher information is cached
-    /// without hitting the network.
-    pub fn get_cached_publishers(&self, name: PackageStr<'_>) -> Vec<PublisherCacheVersion> {
-        let guard = self.state.lock().unwrap();
-        guard
-            .publisher_cache
-            .crates
-            .get(name)
-            .map_or_else(Vec::new, |c| c.versions.clone())
-    }
-
-    /// Look up information about who published each version of the specified
-    /// crates. Versions for each crate are also specified in order to avoid
-    /// hitting the network in the case where the cache already has the relevant
-    /// information.
-    pub async fn get_publishers(
+    /// If `versions` is specified, the cached information will be used if all specified versions
+    /// are already present or if the missing versions are _not_ in the crates.io index and the
+    /// last fetched time is less than `NONINDEX_VERSION_PUBLISHER_REFRESH_DAYS`.
+    ///
+    /// If `versions` is None, the cached information is used _only_ when the last fetched time is
+    /// less than `NONINDEX_VERSION_PUBLISHER_REFRESH_DAYS`.
+    ///
+    /// When this function returns `Ok`, the returned state is guaranteed to have an entry for
+    /// `name` in `publisher_cache.crates`.
+    async fn update_publisher_cache(
         &self,
         network: &Network,
         name: PackageStr<'_>,
-        versions: FastSet<&semver::Version>,
-    ) -> Result<Vec<PublisherCacheVersion>, GetPublishersError> {
+        versions: Option<FastSet<&'_ semver::Version>>,
+    ) -> Result<std::sync::MutexGuard<CacheState>, GetPublishersError> {
```

Comment:
```suggestion
        versions: Option<FastSet<&semver::Version>>,
    ) -> Result<std::sync::MutexGuard<'_, CacheState>, GetPublishersError> {
```

I like to include the lifetime placeholder explicitly.